gadget.1
.TH "Gadget" "1" "2.1.06" "James Begley" ""
.SH "NAME"
.LP 
gadget
.br 
\fBG\fRlobally applicable \fBA\fRrea \fBD\fRisaggregated \fBG\fReneral \fBE\fRcosystem \fBT\fRoolbox
.SH "SYNTAX"
.LP 
gadget [\fIoptions\fP]
.SH "DESCRIPTION"
.LP 
Gadget is a flexible and powerful tool for creating ecosystem models.  The program was developed for modelling marine ecosystems in a fisheries management and biology context, however there is nothing in the program that restricts it to fish, and models have been developed to examine marine mammal populations

.LP 
Gadget can run complicated statistical ecosystem models, which take many features of the ecosystem into account. Gadget works by running an internal model based on many parameters, and then comparing the data from the output of this model to real data to get a goodness\-of\-fit likelihood score.  The parameters can then be adjusted, and the model re\-run, until an optimum is found, which corresponds to the model with the lowest likelihood score.

.LP 
Gadget allows you to include a number of features into your model: One or more species, each of which may be split into multiple stocks; multiple areas with migration between areas; predation between and within species; maturation; reproduction and recruitment; multiple commercial and survey fleets taking catches from the populations.
.SH "OPTIONS"
.LP 
Options for running Gadget:
.TP 
\fB\-l\fR
perform a likelihood (optimising) model run
.TP 
\fB\-s\fR
perform a single (simulation) model run
.TP 
\fB\-n\fR
perform a network run (using paramin)
.TP 
\fB\-v  \-\-version\fR
display version information and exit
.TP 
\fB\-h  \-\-help\fR
display help screen and exit
 
.LP 
Options for specifying the input to Gadget models:
.TP 
\fB\-i <filename>\fR
read model parameters from <filename>
.TP 
\fB\-opt <filename>\fR
read optimising parameters from <filename>
.TP 
\fB\-main <filename>\fR
read model information from <filename> (default "main")
.TP 
\fB\-m <filename>\fR
read other commandline parameters from <filename>
 
.LP 
Options for specifying the output from Gadget models:
.TP 
\fB\-p <filename>\fR
print model parameters to <filename> (default "params.out")
.TP 
\fB\-o <filename>\fR
print likelihood output to <filename>
.TP 
\fB\-print <number>\fR
print \-o output every <number> iterations
.TP 
\fB\-precision <number>\fR
set the precision to <number> in output files
 
.LP 
Options for debugging Gadget models:
.TP 
\fB\-log <filename>\fR
print logging information to <filename>
.TP 
\fB\-printinitial <filename>\fR
print initial model information to <filename>
.TP 
\fB\-printfinal <filename>\fR
print final model information to <filename>
 
.SH "ENVIRONMENT VARIABLES"
.LP 
.TP 
\fBGADGET_WORKING_DIR\fP
Specifies the working directory, used for the output from Gadget (defaults to current directory if not specified)
.TP 
\fBGADGET_DATA_DIR\fP
Specifies the data directory, used for the input to Gadget (defaults to current directory if not specified)
.SH "CONTACT"
.LP 
For help on running Gadget, constructing Gadget models and
understanding the output file an issue on https://github.com/gadget-framework/gadget2 with
the label help-wanted
.LP 
There is also a Gadget user mailing list set up <gadgetusers@hafro.is>
.LP 
Report software bugs at github.com/Hafro/gadget2 
.LP 
Documentation written by James Begley
.SH "SEE ALSO"
.LP 
Gadget web page <https://hafro.github.io/gadget2>


src/gadget.cc
#include "ecosystem.h"
#include "maininfo.h"
#include "runid.h"
#include "gadget.h"
#include "errorhandler.h"
#include "stochasticdata.h"
#include "interrupthandler.h"
#include "global.h"
#ifdef _OPENMP
#include <omp.h>
#endif

Ecosystem* EcoSystem;
StochasticData* data = 0;
Ecosystem** EcoSystems;
MainInfo _main;

volatile int interrupted = 0;
volatile int interrupted_print = 0;


int main(int aNumber, char* const aVector[]) {
  int check = 0;
  //Initialise random number generator with system time [MNAA 02.02.26]
  srand((int)time(NULL));

  //Test to see if the function double lgamma(double) is returning an integer.
  //lgamma is a non-ansi function and on some platforms when compiled with the
  //-ansi flag lgamma returns an integer value. [MNAA&AJ 05.2001]
  assert(lgamma(1.2) != floor(lgamma(1.2)));

  char* workingdir = getenv("GADGET_WORKING_DIR");
  if (workingdir == 0) {
    if ((workingdir = (char*)malloc(LongString)) == NULL)
      handle.logMessage(LOGFAIL, "Error - failed to malloc space for current working directory");
    check = 1;
    if (getcwd(workingdir, LongString) == NULL)
      handle.logMessage(LOGFAIL, "Error - failed to get current working directory");
  }
  if (chdir(workingdir) != 0)
    handle.logMessage(LOGFAIL, "Error - failed to change working directory to", workingdir);

  char* inputdir = getenv("GADGET_DATA_DIR");
  if (inputdir == 0)
    inputdir = workingdir;
  if (chdir(inputdir) != 0)
    handle.logMessage(LOGFAIL, "Error - failed to change input directory to", inputdir);
  if (chdir(workingdir) != 0) //JMB change back to where we were ...
    handle.logMessage(LOGFAIL, "Error - failed to change working directory to", workingdir);

  _main.read(aNumber, aVector);
  _main.checkUsage(inputdir, workingdir);

  if (chdir(inputdir) != 0)
    handle.logMessage(LOGFAIL, "Error - failed to change input directory to", inputdir);
  EcoSystem = new Ecosystem(_main);

#ifdef INTERRUPT_HANDLER
  //JMB dont register interrupt if doing a network run
  if (!(_main.runNetwork()))
    registerInterrupts(&interrupted);
#endif

  if (chdir(workingdir) != 0)
    handle.logMessage(LOGFAIL, "Error - failed to change working directory to", workingdir);
  if ((_main.getPI()).getPrint())
    EcoSystem->writeInitialInformation((_main.getPI()).getOutputFile());

  if (_main.runStochastic()) {
    if (_main.runNetwork()) {
#ifdef GADGET_NETWORK //to help compiling when pvm libraries are unavailable
      EcoSystem->Initialise();
      data = new StochasticData();
      while (data->getDataFromNetwork()) {
        EcoSystem->Update(data);
        EcoSystem->Simulate(_main.runPrint());
        data->sendDataToNetwork(EcoSystem->getLikelihood());
        data->readNextLineFromNetwork();
      }
      delete data;
#endif

    } else if (_main.getInitialParamGiven()) {
      if (chdir(inputdir) != 0) //JMB need to change back to inputdir to read the file
        handle.logMessage(LOGFAIL, "Error - failed to change input directory to", inputdir);
      data = new StochasticData(_main.getInitialParamFile());
      if (chdir(workingdir) != 0)
        handle.logMessage(LOGFAIL, "Error - failed to change working directory to", workingdir);

      EcoSystem->Update(data);
      EcoSystem->checkBounds();

      EcoSystem->Initialise();
      if (_main.printInitial()) {
        EcoSystem->Reset();  //JMB only need to call reset() before the print commands
        EcoSystem->writeStatus(_main.getPrintInitialFile());
      }

      EcoSystem->Simulate(_main.runPrint());
      if ((_main.getPI()).getPrint())
        EcoSystem->writeValues();

      while (data->isDataLeft()) {
        data->readNextLine();
        EcoSystem->Update(data);
        //EcoSystem->checkBounds();
        EcoSystem->Simulate(_main.runPrint());
        if ((_main.getPI()).getPrint())
          EcoSystem->writeValues();
      }
      delete data;

    } else {
      if (EcoSystem->numVariables() != 0)
        handle.logMessage(LOGWARN, "Warning - no parameter input file given, using default values");

      EcoSystem->Initialise();
      if (_main.printInitial()) {
        EcoSystem->Reset();  //JMB only need to call reset() before the print commands
        EcoSystem->writeStatus(_main.getPrintInitialFile());
      }

      EcoSystem->Simulate(_main.runPrint());
      if ((_main.getPI()).getPrint())
        EcoSystem->writeValues();
    }

  } else if (_main.runOptimise()) {
	  if (EcoSystem->numVariables() == 0)
		  handle.logMessage(LOGFAIL, "Error - no parameters can be optimised");

	  if (_main.getInitialParamGiven()) {
		  if (chdir(inputdir) != 0) //JMB need to change back to inputdir to read the file
			handle.logMessage(LOGFAIL, "Error - failed to change input directory to", inputdir);
		  data = new StochasticData(_main.getInitialParamFile());
		  if (chdir(workingdir) != 0)
			handle.logMessage(LOGFAIL, "Error - failed to change working directory to", workingdir);

		  EcoSystem->Update(data);
		  EcoSystem->checkBounds();
		} else
		  handle.logMessage(LOGFAIL, "Error - no parameter input file specified");

	  EcoSystem->Initialise();
	  if (_main.printInitial()) {
		  EcoSystem->Reset();  //JMB only need to call reset() before the print commands
		  EcoSystem->writeStatus(_main.getPrintInitialFile());
	  	  }

#ifdef _OPENMP
          if (_main.runParallel()){
	  	int numThr = omp_get_max_threads ( );
		handle.logMessage(LOGINFO, "Info - initialise ", numThr," EcoSystems ");
	  	EcoSystems = new Ecosystem*[numThr];
	  	int i;
#pragma omp parallel for shared(EcoSystems,_main,data,handle)
	  	for (i=0; i<numThr; i++)
	  	{
		        handle.logMessage(LOGINFO, "Info - Ecosystem ", i);
	                if (_main.getInitialParamGiven()) {
		  	    EcoSystems[i] = new Ecosystem(_main);
		  	    EcoSystems[i]->Update(data);
		            EcoSystems[i]->checkBounds();
			}
		  	EcoSystems[i]->Initialise();
	  	}
	  	}
#endif
	  EcoSystem->Optimise();
	  delete data;
	  if (_main.getForcePrint())
		  EcoSystem->Simulate(_main.getForcePrint());
  }

  handle.logMessage(LOGMESSAGE, "");  //write blank line to log file
  if (_main.printFinal() && !(_main.runNetwork()))
    EcoSystem->writeStatus(_main.getPrintFinalFile());

  //JMB print final values of parameters
  if (!(_main.runNetwork()))
    EcoSystem->writeParams((_main.getPI()).getParamOutFile(), (_main.getPI()).getPrecision());

  if (check)
    free(workingdir);

  delete EcoSystem;
  handle.logFinish();
  return EXIT_SUCCESS;
}
